presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedTruncatedHistogram.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.BlockBuilderStatus;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.type.MapType;
+import com.facebook.presto.util.array.IntBigArray;
+import com.facebook.presto.util.array.LongBigArray;
+import com.google.common.primitives.Ints;
+import io.airlift.units.DataSize;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.ExceededMemoryLimitException.exceededLocalLimit;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+import static it.unimi.dsi.fastutil.HashCommon.murmurHash3;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A histogram that periodically truncates infrequent items and overflows their counters onto other items.
+ */
+public class TypedTruncatedHistogram
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(TypedTruncatedHistogram.class).instanceSize();
+
+    private static final float FILL_RATIO = 0.75f;
+    private static final float COMPACT_RATIO = 0.5f;
+    private static final long FOUR_MEGABYTES = new DataSize(4, MEGABYTE).toBytes();
+
+    private int maxFill;
+    private int mask;
+    private final int compactLimit;
+
+    private final Type type;
+
+    private BlockBuilder values;
+    private IntBigArray hashPositions;
+    private final LongBigArray counts;
```

Comment:
We put all `final` fields before the mutable ones, so move this up as well as `compactLimit`